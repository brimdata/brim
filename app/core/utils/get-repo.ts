import packageJSON from "./package-json"

export async function getRepo(url?: string) {
  url = url || (await packageJSON()).repository
  return new URL(url).pathname.slice(1)
}
