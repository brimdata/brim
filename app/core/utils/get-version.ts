import {remote} from "electron"

export function getVersion() {
  return remote.app.getVersion()
}
