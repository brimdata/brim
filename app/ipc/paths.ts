import {app} from "electron"
import os from "os"
import {join} from "path"
import {createClient} from "src/pkg/electron-ipc-service"

const exe = (name) => (os.platform() === "win32" ? `${name}.exe` : name)

class Paths {
  brimcap() {
    return join(
      app.getAppPath().replace("app.asar", "app.asar.unpacked"),
      "zdeps",
      exe("brimcap")
    )
  }
}

export const paths = new Paths()
export const pathsClient = createClient(Paths)
