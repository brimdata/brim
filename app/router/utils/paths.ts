export const workspacePath = (id) => `/workspaces/${id}`
export const workspacesPath = () => `/workspaces`
export const lakePath = (id, workspaceId) =>
  `/workspaces/${workspaceId}/lakes/${id}`
