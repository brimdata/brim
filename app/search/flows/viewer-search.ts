import {ANALYTIC_MAX_RESULTS, PER_PAGE} from "src/js/flows/config"
import {search, SearchResult} from "src/js/flows/search/mod"
import ErrorFactory from "src/js/models/ErrorFactory"
import Columns from "src/js/state/Columns"
import Current from "src/js/state/Current"
import Notice from "src/js/state/Notice"
import SearchBar from "src/js/state/SearchBar"
import Tabs from "src/js/state/Tabs"
import {Thunk} from "src/js/state/types"
import Viewer from "src/js/state/Viewer"
import {SchemaMap} from "src/js/state/Viewer/types"
import {zed} from "@brimdata/zealot"
import {SearchResponse} from "../../../src/js/flows/search/response"

type Args = {
  query: string
  from: Date
  to: Date
  keep?: boolean
  append?: boolean
}

const id = "Table"

export function viewerSearch(args: Args): Thunk<Promise<SearchResult>> {
  return (dispatch, getState) => {
    const {query, from, to, keep, append} = args
    const tabId = Tabs.getActive(getState())
    const poolId = Current.mustGetPool(getState()).id
    const {response, promise} = dispatch(
      search({id, query, from, to, poolId, initial: !append})
    )
    dispatch(handle(response, tabId, keep, append))
    return promise.catch((e) => e)
  }
}

function handle(
  response: SearchResponse,
  tabId: string,
  keep = false,
  append = false
): Thunk {
  return function(dispatch, getState) {
    let allColumns: SchemaMap = {}
    let allRecords: zed.Record[] = []
    let count = 0

    // preserve searchKey in case of clear
    const currentSearchKey = Viewer.getSearchKey(getState())
    if (!keep && !append) {
      dispatch(Viewer.clear(tabId))
      dispatch(Viewer.setSearchKey(tabId, currentSearchKey))
    }

    dispatch(Viewer.setStatus(tabId, "FETCHING"))
    dispatch(Viewer.setEndStatus(tabId, "FETCHING"))

    response
      .status((status) => dispatch(Viewer.setStatus(tabId, status)))
      .chan(0, ({rows, schemas}) => {
        count = rows.length

        if (keep) {
          allColumns = schemas
          allRecords = rows
          return
        }

        if (append) {
          dispatch(Viewer.appendRecords(tabId, rows))
        } else {
          dispatch(Viewer.setRecords(tabId, rows))
        }
        dispatch(Viewer.updateColumns(tabId, schemas))
        dispatch(Columns.touch(schemas))
      })
      .warning((warning) => dispatch(SearchBar.errorSearchBarParse(warning)))
      .error((error) => {
        dispatch(Notice.set(ErrorFactory.create(error)))
      })
      .end(() => {
        if (keep) {
          dispatch(Viewer.clear(tabId))
          dispatch(Viewer.setSearchKey(tabId, currentSearchKey))
          dispatch(Viewer.setRecords(tabId, allRecords))
          dispatch(Viewer.setColumns(tabId, allColumns))
          dispatch(Columns.touch(allColumns))
        }
        dispatch(Viewer.setEndStatus(tabId, endStatus(count)))
      })
  }
}

function endStatus(count) {
  if (count === PER_PAGE) return "INCOMPLETE"
  if (count === ANALYTIC_MAX_RESULTS) return "LIMIT"
  return "COMPLETE"
}
