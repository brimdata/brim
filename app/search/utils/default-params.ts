import {BrimSpace} from "src/js/brim"
import {SpanArgs} from "src/js/state/Search/types"

export function mergeDefaultSpanArgs(
  spanArgs: Partial<SpanArgs>,
  space: BrimSpace
): SpanArgs {
  const [from, to] = spanArgs
  if (!from && !to) return space.everythingSpan()
  if (!from) return [space.minTs(), to]
  if (!to) return [from, space.maxTs()]
  return [from, to]
}
