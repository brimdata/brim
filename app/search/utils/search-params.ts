import {TimeArg, SpanArgs} from "src/js/state/Search/types"

type Args = {
  program: string
  spanArgs: SpanArgs
  pins: string[]
  spanArgsFocus: SpanArgs
}

export const encodeSearchParams = ({
  program,
  pins,
  spanArgs,
  spanArgsFocus
}: Partial<Args>) => {
  const p = new URLSearchParams()
  if (program) p.append("q", program)
  encodeSpan(p, spanArgs, "from", "to")
  encodeSpan(p, spanArgsFocus, "focusFrom", "focusTo")
  encodePins(p, pins || [])
  return p.toString()
}

type DecodedSearchParams = {
  program: string
  pins: string[]
  spanArgs: Partial<SpanArgs>
  spanArgsFocus: Partial<SpanArgs>
}

export const decodeSearchParams = (path: string): DecodedSearchParams => {
  const url = new URLSearchParams(path)
  return {
    program: url.get("q"),
    spanArgs: [url.get("from"), url.get("to")].map(decodeSpanArg) as SpanArgs,
    spanArgsFocus: [url.get("focusFrom"), url.get("focusTo")].map(
      decodeSpanArg
    ) as SpanArgs,
    pins: decodePins(url)
  }
}

const encodeSpan = (params, span, from, to) => {
  if (span && span[0]) params.append(from, encodeSpanArg(span[0]))
  if (span && span[1]) params.append(to, encodeSpanArg(span[1]))
}

const encodeSpanArg = (arg) => {
  return typeof arg === "string" ? arg : `${arg.sec}.${arg.ns}`
}

const decodeSpanArg = (arg): TimeArg | null => {
  if (!arg) return null
  if (/^\d+\.\d+$/.test(arg)) {
    const [sec, ns] = arg.split(".").map((i) => parseInt(i))
    return {sec, ns}
  } else {
    return arg
  }
}

const pinKey = (i) => `p${i}`

const encodePins = (params, pins) => {
  for (let i = 0; i < pins.length; ++i) {
    params.append(pinKey(i), pins[i])
  }
}

const decodePins = (params) => {
  const pins = []
  for (let i = 0; params.has(pinKey(i)); i++) pins.push(params.get(pinKey(i)))

  return pins
}
