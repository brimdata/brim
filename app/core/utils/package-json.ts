import {pathsClient} from "app/ipc/paths"
import {readJSON} from "fs-extra"
import {join} from "path"

let cache

export default async function packageJSON() {
  if (cache) return cache
  const root = await pathsClient.root()
  const filePath = join(root, "package.json")
  return (cache = readJSON(filePath))
}
