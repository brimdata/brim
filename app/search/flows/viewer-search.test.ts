import tabHistory from "app/router/tab-history"
import {lakePath} from "app/router/utils/paths"
import Columns from "src/js/state/Columns"
import SearchBar from "src/js/state/SearchBar"
import Pools from "src/js/state/Pools"
import Viewer from "src/js/state/Viewer"
import Workspaces from "src/js/state/Workspaces"
import fixtures from "test/unit/fixtures"
import initTestStore, {TestStore} from "test/unit/helpers/initTestStore"
import {createRecord} from "test/shared/factories/zed-factory"
import {useResponse} from "test/shared/responses"
import {createZealotMock, ZealotMock} from "zealot-old"
import {viewerSearch} from "./viewer-search"

const dnsResp = useResponse("dns")
const pool = fixtures("pool1")
const warningResp = useResponse("searchWarning")

let dispatch, select
let zealot: ZealotMock
let store: TestStore
beforeEach(() => {
  zealot = createZealotMock()
  store = initTestStore(zealot.zealot)
  dispatch = store.dispatch
  select = (s: any) => s(store.getState())

  store.dispatchAll([
    Workspaces.add({
      host: "testHost",
      id: "1",
      name: "testName",
      port: "9867",
      authType: "none"
    }),
    Pools.setDetail("1", pool)
  ])
  store.dispatch(tabHistory.push(lakePath(pool.id, "1")))
})

const getQueryCallChecker = () => {
  let callCount = 0
  return async (query: string, expectAnnotation = false) => {
    zealot.stubStream("query", dnsResp)
    await dispatch(
      viewerSearch({
        query,
        from,
        to
      })
    )
    callCount++
    const calls = zealot.calls("query")
    expect(calls.length).toBe(callCount)
    const expected = expectAnnotation
      ? `from '1' | ts >= ${from.toISOString()} | ts <= 1970-01-01T00:00:00.001Z | ${query}`
      : query
    expect(calls[callCount - 1].args).toEqual(expected)
  }
}

const from = new Date()
const to = new Date(1)
const submit = () =>
  dispatch(
    viewerSearch({
      query: "dns query | head 500",
      from,
      to
    })
  )

describe("a normal response", () => {
  beforeEach(() => {
    zealot.stubStream("query", dnsResp)
  })

  test("zealot gets the request", async () => {
    const checkQueryCall = getQueryCallChecker()
    await checkQueryCall("dns query | head 500", true)
  })

  test("zealot does not annotate requests beginning with variations of 'from'", async () => {
    const checkQueryCall = getQueryCallChecker()
    await checkQueryCall("from 'test' | test")
    await checkQueryCall("from('test') | test")
    await checkQueryCall("from ('test') | test")
    await checkQueryCall("from    ('test') | test")
  })

  test("the table gets populated", async () => {
    await submit()
    expect(select(Viewer.getViewerRecords).length).toBe(2)
  })

  test("the table gets cleared", async () => {
    dispatch(Viewer.setRecords(undefined, [createRecord({clear: "me"})]))
    await submit()
    expect(select(Viewer.getViewerRecords)[0]).not.toEqual([
      {name: "clear", type: "string", value: "me"}
    ])
  })

  test("the table status updates", async () => {
    const promise = submit()
    expect(select(Viewer.getStatus)).toBe("FETCHING")
    expect(select(Viewer.getEndStatus)).toBe("FETCHING")
    await promise
    expect(select(Viewer.getStatus)).toBe("SUCCESS")
    expect(select(Viewer.getEndStatus)).toBe("COMPLETE")
  })

  test("sets the viewer columns", async () => {
    await submit()
    expect(select(Viewer.getColumns)).toMatchSnapshot()
  })

  test("sets the Columns columns", async () => {
    await submit()
    expect(select(Columns.getColumns)).toMatchSnapshot()
  })
})

test("a response with a warning", async () => {
  zealot.stubStream("query", warningResp)
  await submit()
  expect(select(SearchBar.getSearchBarError)).toBe(
    "Sort field boo not present in input"
  )
})
