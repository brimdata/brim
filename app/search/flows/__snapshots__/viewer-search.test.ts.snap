// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a normal response sets the Columns columns 1`] = `
Object {
  "7247fae6fe9d9e9087c6394e7aba3004": Object {
    "AA:bool": Object {
      "isVisible": true,
    },
    "RA:bool": Object {
      "isVisible": true,
    },
    "RD:bool": Object {
      "isVisible": true,
    },
    "TC:bool": Object {
      "isVisible": true,
    },
    "TTLs:array": Object {
      "isVisible": true,
    },
    "Z:count": Object {
      "isVisible": true,
    },
    "_path:string": Object {
      "isVisible": true,
    },
    "answers:array": Object {
      "isVisible": true,
    },
    "id.orig_h:addr": Object {
      "isVisible": true,
    },
    "id.orig_p:port": Object {
      "isVisible": true,
    },
    "id.resp_h:addr": Object {
      "isVisible": true,
    },
    "id.resp_p:port": Object {
      "isVisible": true,
    },
    "proto:enum": Object {
      "isVisible": true,
    },
    "qclass:count": Object {
      "isVisible": true,
    },
    "qclass_name:string": Object {
      "isVisible": true,
    },
    "qtype:count": Object {
      "isVisible": true,
    },
    "qtype_name:string": Object {
      "isVisible": true,
    },
    "query:string": Object {
      "isVisible": true,
    },
    "rcode:count": Object {
      "isVisible": true,
    },
    "rcode_name:string": Object {
      "isVisible": true,
    },
    "rejected:bool": Object {
      "isVisible": true,
    },
    "rtt:interval": Object {
      "isVisible": true,
    },
    "trans_id:count": Object {
      "isVisible": true,
    },
    "ts:time": Object {
      "isVisible": true,
    },
    "uid:string": Object {
      "isVisible": true,
    },
  },
}
`;

exports[`a normal response sets the viewer columns 1`] = `
Object {
  "7247fae6fe9d9e9087c6394e7aba3004": Schema {
    "columns": Array [
      Object {
        "name": "_path",
        "type": "string",
      },
      Object {
        "name": "ts",
        "type": "time",
      },
      Object {
        "name": "uid",
        "type": "string",
      },
      Object {
        "name": "id",
        "of": Array [
          Object {
            "name": "orig_h",
            "type": "addr",
          },
          Object {
            "name": "orig_p",
            "type": "port",
          },
          Object {
            "name": "resp_h",
            "type": "addr",
          },
          Object {
            "name": "resp_p",
            "type": "port",
          },
        ],
        "type": "record",
      },
      Object {
        "name": "proto",
        "type": "enum",
      },
      Object {
        "name": "trans_id",
        "type": "count",
      },
      Object {
        "name": "rtt",
        "type": "interval",
      },
      Object {
        "name": "query",
        "type": "string",
      },
      Object {
        "name": "qclass",
        "type": "count",
      },
      Object {
        "name": "qclass_name",
        "type": "string",
      },
      Object {
        "name": "qtype",
        "type": "count",
      },
      Object {
        "name": "qtype_name",
        "type": "string",
      },
      Object {
        "name": "rcode",
        "type": "count",
      },
      Object {
        "name": "rcode_name",
        "type": "string",
      },
      Object {
        "name": "AA",
        "type": "bool",
      },
      Object {
        "name": "TC",
        "type": "bool",
      },
      Object {
        "name": "RD",
        "type": "bool",
      },
      Object {
        "name": "RA",
        "type": "bool",
      },
      Object {
        "name": "Z",
        "type": "count",
      },
      Object {
        "name": "answers",
        "of": "string",
        "type": "array",
      },
      Object {
        "name": "TTLs",
        "of": "interval",
        "type": "array",
      },
      Object {
        "name": "rejected",
        "type": "bool",
      },
    ],
  },
}
`;
